Sources/NIOKit/ConnectionPool.swift

```diff
@@ -0,0 +1,169 @@
+    ///
+    /// - parameters:
+    ///     - closure: Callback that accepts the pooled connection.
+    /// - returns: A future containing the result of the closure.
+    public func withConnection<Result>(_ closure: @escaping (Source.Connection) -> EventLoopFuture<Result>) -> EventLoopFuture<Result> {
+        return requestConnection().flatMap { conn in
+            return closure(conn).map { res in
+                self.releaseConnection(conn)
+                return res
+            }.flatMapErrorThrowing { error in
+                self.releaseConnection(conn)
+                throw error
+            }
+        }
+    }
+    
+    /// Requests a pooled connection.
+    ///
+    /// The connection returned by this method should be released when you are finished using it.
+    ///
+    ///     let conn = try pool.requestConnection().wait()
+    ///     defer { pool.releaseConnection(conn) }
+    ///     // use the connection
+    ///
+    /// See `withConnection(_:)` for a callback-based method that automatically releases the connection.
+    ///
+    /// - returns: A future containing the requested connection.
+    public func requestConnection() -> EventLoopFuture<Source.Connection> {
+        if let conn = self.available.popLast() {
+            // check if it is still open
+            if !conn.isClosed {
+                // connection is still open, we can return it directly
+                return self.source.eventLoop.makeSucceededFuture(conn)
+            } else {
+                // connection is closed, we need to replace it
+                return self.source.makeConnection()
+            }
+        } else if self.activeConnections < config.maxConnections  {
+            // all connections are busy, but we have room to open a new connection!
+            self.activeConnections += 1
+            
+            // make the new connection
+            return self.source.makeConnection()
+        } else {
+            // connections are exhausted, we must wait for one to be returned
+            let promise = self.source.eventLoop.makePromise(of: Source.Connection.self)
+            self.waiters.append(promise)
+            return promise.futureResult
+        }
+    }
+    
+    /// Releases a connection back to the pool. Use with `requestConnection()`.
+    ///
+    ///     let conn = try pool.requestConnection().wait()
+    ///     defer { pool.releaseConnection(conn) }
+    ///     // use the connection
+    ///
+    /// - parameters:
+    ///     - conn: Connection to release back to the pool.
+    public func releaseConnection(_ conn: Source.Connection) {
+        // get the active connection for this connection
+        self.available.append(conn)
+        
+        // now that we know a new connection is available, we should
+        // take this chance to fulfill one of the waiters
+        if let waiter = self.waiters.popLast() {
+            self.requestConnection().cascade(to: waiter)
```

Comment:
I'm not quite sure what this does, some explanation would be nice 😄

Reply:
The two lines of comment above the `if` should be trying to explain this. What part about it is unclear?

Reply:
🤔 Firstoff, I misread.

Second off, would this not have to be pop first? Otherwise you'd have this right?

```swift
// Conns are full
try pool.requestConnection(conn) {
    print("A")
}

try pool.requestConnection(conn) {
    print("B")
}
```
Now if we release some connections, because of using `popLast()` it'll output
```
B
A
```
Right?

Reply:
@MrLotU hmm good point. Popping from the start of an array is much more expensive though. I should look into using a different data type, maybe NIO's circular buffer?

Devil's advocate: I'm not sure this makes the implementation "incorrect" though. 🤔 As long as each request for a connection eventually gets it, the rest should be an implementation detail right?

Reply:
It does 100% not make it incorrect, it'd just mean that hypothetically speaking, all conns are occupied, you add something that's "important" and than start spamming new requests like a maniac, your first thing will never be hit.

Is appending at the front maybe a solution? Not sure how expensive that is in comparison to removing from the back/front

Reply:
As for circular buffer, I'm not exactly sure how that works :P

Reply:
The two methods used currently are `O(1)` (at least on average).

[`Array.popLast()`](https://developer.apple.com/documentation/swift/array/1539777-poplast) is complexity `O(1)`
[`Array.append(_:)`](https://developer.apple.com/documentation/swift/array/3126937-append) is complexity `O(1)` on average.

The methods required for this change are both `O(n)`

[`Array.insert(_:at)`](https://developer.apple.com/documentation/swift/array/3126951-insert) is complexity `O(n)`.
[`Array.remove(at:)`](https://developer.apple.com/documentation/swift/array/1641390-remove) is complexity `O(n)`.

Reply:
> As for circular buffer, I'm not exactly sure how that works :P

A circular buffers append _and_ prepend are `O(1)` on average.

Reply:
@tanner0101 I see!

I don't think it's a must fix right now, but something that should be fixed in the future, since it's not breaking or anything, just in the case where you have a LOT of ongoing connections and appends

Reply:
Sounds good. I'll open an issue for this once merged.